### PR TITLE
Added  task retry model feature to task properties.

### PR DESCRIPTION
**Pull request by shital-orchestral**
Added **Retry Feature** as a new field under task properties.

### CODE REVIEW

**Review by m4dcoder — CHANGES_REQUESTED** [review]
@shital-orchestral A few changes

1. Can you change the default value for the `when` field to `<% failed() %>`? The default expression right now is not a valid expression for the context of retry.
2. Can you change the text for delay to `delay (seconds)`?

File: modules/st2-auto-form/style.css [hunk 1]
```diff
@@ -480,7 +480,22 @@
...
+   margin-left:4px;
+ }
+}
 
-
-
-
+.count-title{
+ &:hover:after {
+   content:"enter integer value";
```
Comment (m4dcoder) [inline modules/st2-auto-form/style.css:492]: Please change text to `enter expression or integer`. Can we change it to a different color (i.e. gray)?  Can it be a smaller font size?

File: modules/st2-auto-form/style.css [hunk 2]
```diff
@@ -480,7 +480,22 @@
...
+   margin-left:4px;
+ }
+}
+.delay-title{
+ &:hover:after {
+   content:"enter integer value";
```
Comment (m4dcoder) [inline modules/st2-auto-form/style.css:498]: Please change text to `enter expression or integer`. Can we change it to a different color (i.e. gray)?  Can it be a smaller font size?

File: modules/st2-auto-form/style.css [hunk 3]
```diff
@@ -480,7 +480,22 @@
     font-family: Garamond, Georgia, Times New Roman, serif !important;;
   }
 }
+.when-title{
+ &:hover:after {
+   content:"enter expression";
+   margin-left:4px;
+ }
```
Comment (m4dcoder) [inline modules/st2-auto-form/style.css:487]: Can we change it to a different color (i.e. gray)?  Can it be a smaller font size?

File: modules/st2flow-model/schemas/orquesta.json [hunk 4]
```diff
@@ -327,4 +364,4 @@
       "type": "string"
     }
   }
-}
+}
```
Comment (m4dcoder) [inline modules/st2flow-model/schemas/orquesta.json:367]: Add a newline?

File: modules/st2-auto-form/style.css [hunk 5]
```diff
@@ -480,7 +480,22 @@
...
+   content:"enter integer value";
+   margin-left:4px;
+ }
+}
+.delay-title{
+ &:hover:after {
+   content:"enter integer value";
+   margin-left:4px;
+ }
+}
```
Comment (m4dcoder) [inline modules/st2-auto-form/style.css:501]: Add a newline?

File: modules/st2flow-details/orquesta-properties.js [hunk 6]
```diff
@@ -100,6 +100,17 @@ export default class OrquestaTransition extends Component<TransitionProps, {}> {
           )
         }
       </Property>,
+      <Property key="retry" name="Retry" description="Define the retry condition for the task execution." value={!!task.retry} onChange={value => this.handleTaskProperty('retry', value ? { when: '<% failed() %>' } : false)}>
+        {
+          task.retry && (
+            <div className={this.style.propertyChild}>
+              <StringField name="when" value={task.retry.when} className="when-title" onChange={value => this.handleTaskProperty([ 'retry', 'when' ], value)} />
+              <StringField name="count" value={task.retry.count} className="count-title" onChange={value => this.handleTaskProperty([ 'retry', 'count' ], value)} />
+              <StringField name="delay" value={task.retry.delay} className="delay-title" onChange={value => this.handleTaskProperty([ 'retry', 'delay(seconds)'], value)} />
```
Comment (m4dcoder) [inline modules/st2flow-details/orquesta-properties.js:109]: Can we change field name to `delay (seconds)`?

**Review by m4dcoder — CHANGES_REQUESTED** [review]
@shital-orchestral Can we make the help text appears when we hover on the input box and not hover on the field name?

File: modules/st2flow-details/orquesta-properties.js [hunk 7]
```diff
@@ -100,6 +100,17 @@ export default class OrquestaTransition extends Component<TransitionProps, {}> {
           )
         }
       </Property>,
+      <Property key="retry" name="Retry" description="Define the retry condition for the task execution." value={!!task.retry} onChange={value => this.handleTaskProperty('retry', value ? { when: '<% failed() %>' } : false)}>
+        {
+          task.retry && (
+            <div className={this.style.propertyChild}>
+              <StringField name="when" value={task.retry.when} className="when-title" onChange={value => this.handleTaskProperty([ 'retry', 'when' ], value)} />
+              <StringField name="count" value={task.retry.count} className="count-title" onChange={value => this.handleTaskProperty([ 'retry', 'count' ], value)}  spec={{'default':'enter expression or integer'}} />
+              <StringField name="delay(seconds)" value={task.retry.delay} className="delay-title" onChange={value => this.handleTaskProperty([ 'retry', 'delay'], value)} spec={{'default':'enter expression or integer'}} />
```
Comment (m4dcoder) [inline modules/st2flow-details/orquesta-properties.js:109]: Can you put a space between delay and (seconds) so it's `delay (seconds)`?

**Review by m4dcoder — CHANGES_REQUESTED** [review]
@shital-orchestral What happen to the help text?

**Review by m4dcoder — APPROVED** [review]

**Comment by amanda11** [comment]
@m4dcoder @shital-orchestral LGTM. I found no problems with this PR, you can set negative numbers, but as per the other PR I think that is a limitation we have with other fields, and where we need to allow expression support not worth trying to resolve.

I found a bug when testing it but confirmed that that problem also exists on the master branch - so not to do with this PR. But it's an interesting bug as there are problems with Save, Execute, Save if your workflow has an input parameter: https://github.com/StackStorm/st2web/issues/862. I have confirmed that this was NOT a problem with 3.2

**Comment by m4dcoder** [comment]
@amanda11 Here's a screen shot of the UI with an example where count is -100 and a validation error at the bottom of the canvas.

Screen Shot 2021-03-29 at 10 57 21 AM

**Comment by amanda11** [comment]
Looks good - thanks for changes

**Review by amanda11 — APPROVED** [review]
LGTM

**Pull request merged**